server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/FilterToSearchQuery.java

```diff
@@ -0,0 +1,110 @@
+
+import org.apache.james.jmap.model.Filter;
+import org.apache.james.jmap.model.FilterCondition;
+import org.apache.james.jmap.model.FilterOperator;
+import org.apache.james.mailbox.model.SearchQuery;
+import org.apache.james.mailbox.model.SearchQuery.AddressType;
+import org.apache.james.mailbox.model.SearchQuery.Criterion;
+
+public class FilterToSearchQuery {
+
+    public SearchQuery map(Filter filter) {
+        if (filter instanceof FilterCondition) {
+            return mapCondition((FilterCondition) filter);
+        }
+        if (filter instanceof FilterOperator) {
+            return mapOperator((FilterOperator) filter);
+        }
+        throw new RuntimeException("Unknown filter");
+    }
+
+    private SearchQuery mapCondition(FilterCondition filter) {
+        SearchQuery searchQuery = new SearchQuery();
+        if (filter.getText().isPresent()) {
+            String text = filter.getText().get();
+            searchQuery.andCriteria(
+                    SearchQuery.or(SearchQuery.address(AddressType.From, text),
+                            SearchQuery.or(SearchQuery.address(AddressType.To, text),
+                                    SearchQuery.or(SearchQuery.address(AddressType.Cc, text),
+                                            SearchQuery.or(SearchQuery.address(AddressType.Bcc, text),
+                                                    SearchQuery.or(SearchQuery.headerContains("Subject", text),
+                                                            SearchQuery.bodyContains(text))
+                                                    )
+                                            )
+                                    )
+                            )
+                    );
+        }
+        if (filter.getFrom().isPresent()) {
+            searchQuery.andCriteria(SearchQuery.address(AddressType.From, filter.getFrom().get()));
+        }
+        if (filter.getTo().isPresent()) {
+            searchQuery.andCriteria(SearchQuery.address(AddressType.To, filter.getTo().get()));
+        }
+        if (filter.getCc().isPresent()) {
+            searchQuery.andCriteria(SearchQuery.address(AddressType.Cc, filter.getCc().get()));
+        }
+        if (filter.getBcc().isPresent()) {
+            searchQuery.andCriteria(SearchQuery.address(AddressType.Bcc, filter.getBcc().get()));
+        }
+        if (filter.getSubject().isPresent()) {
+            searchQuery.andCriteria(SearchQuery.headerContains("Subject", filter.getSubject().get()));
+        }
+        if (filter.getBody().isPresent()) {
+            searchQuery.andCriteria(SearchQuery.bodyContains(filter.getBody().get()));
+        }
+        return searchQuery;
+    }
+
+    private SearchQuery mapOperator(FilterOperator filter) {
+        SearchQuery searchQuery = new SearchQuery();
+        Consumer<Criterion> andCriteria = f -> searchQuery.andCriteria(f);
+        Stream<List<Criterion>> mappedConditions = filter.getConditions().stream()
```

Comment:
I can't follow the logic.
First `.map(this::map)`is not very helpful
Second, it looks like you build SearchQueries but right after you just get a List of Criterias.
Third, what the hell is a Stream<List<>> useful for ? Isn't what flapmap is supposed to manage ?
You need a bit of work on that method